tests/main/views/test_audits.py

```diff
@@ -50,8 +54,204 @@ def add_audit_events_with_db_object(self):
+                ae.created_at = created_at
+                ae.acknowledged_at = acknowledged_at
+                ae.acknowledged = bool(acknowledged_at)
+                db.session.add(ae)
+                audit_events.append(ae)
+
+            db.session.commit()
+            # make a note of the ids that were given to these events, or rather the order they were generated
+            audit_event_id_lookup = {ae.id: i for i, ae in enumerate(audit_events)}
+            assert AuditEvent.query.count() == len(service_audit_event_params)+len(supplier_audit_event_params)
+
+            return audit_event_id_lookup
+
 
 class TestAuditEvents(BaseTestAuditEvents):
+    @pytest.mark.parametrize(
+        "service_audit_event_params,supplier_audit_event_params,req_params,expected_resp_events",
+        # where we refer to "id"s in the expected_response_params, because we can't be too sure about the *actual* ids
+        # given to objects, we're using 0-based notional "ids" mased on the order the audit events were inserted into
+        # the db. service_audit_event_params events are inserted in the order given, followed by the
+        # supplier_audit_event_params events. so if we had 5 service_audit_event_params and 2
+        # supplier_audit_event_params, "5" would refer to the audit event created by the first-listed
+        # supplier_audit_event_params.
+        # similarly, where supplier and service "id"s are referred to, the "id"s we're referring to are normalized
+        # pseudo-ids from 0-4 inclusive
+        chain.from_iterable(
+            ((serv_aeps, supp_aeps, req_params, expected_resp_events) for req_params, expected_resp_events in req_cases)
+            for serv_aeps, supp_aeps, req_cases in (
+                (
+                    (  # service_audit_event_params
+                        (0, AuditTypes.update_service, datetime(2010, 6, 6), None,),
+                        (0, AuditTypes.update_service, datetime(2010, 6, 7), None,),
+                        (4, AuditTypes.update_service, datetime(2010, 6, 2), None,),
+                    ),
+                    (  # supplier_audit_event_params
+                        (0, AuditTypes.supplier_update, datetime(2010, 6, 6), None,),
+                    ),
+                    (  # and now a series of req_cases - pairs of (req_params, expected_resp_events) to test against
+                       # the above db scenario. these get flattened out into concrete test scenarios by
+                       # chain.from_iterable above before they reach pytest's parametrization
+                        (
+                            {"earliest_for_each_object": "true", "latest_first": "true"},
+                            (3, 0, 2,),
+                        ),
+                        (
+                            {"earliest_for_each_object": "true", "latest_first": "false"},
+                            (2, 0, 3,),
```

Comment:
Well... it's the id. Having a deterministic output becomes particularly useful if you want a query that, when you ask for the reverse ordering, you actually get something that's the reverse ordering, not *mostly* the reverse ordering (Imagine you're wanting to be able to ask for e.g. "the events after number 384827, which I've already seen in the last page I fetched"). Generally people tend to use an API in a way that expects the ordering to be at least stable. If you want I'll dig through the app and find some places where non-deterministic ordering could produce subtly unexpected results.

And if we agree that it's a property that we desire, then it's something that should be tested for surely...?